### PR TITLE
Allow support for `urls` parameter for `generateSprite()` and `multi()`

### PR DESCRIPTION
**Pull request by nitzanj**
This PR adds support for a new parameter called `urls` in two `Uploader` methods: `multi()` and `generateSprite()`.

Both are aggregate methods - they need a set of assets to work on. Until now this set was defined by a `tag` parameter, and now support for a list of urls was added.

Note: you can't omit both 'urls` and 'tag` - one is required.

The second feature added in this PR is the ability to generate a download-url for a sprite or a multi request.

### CODE REVIEW

**Comment by asisayag2** [comment]
@nitzanj , can you add a description of what is supposed to be implemented here?

**Comment by nitzanj** [comment]
> @nitzanj , can you add a description of what is supposed to be implemented here?

@asisayag2 done.

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java [hunk 1]
```diff
@@ -282,13 +292,28 @@ public Map generateSprite(String tag, Map options) throws IOException {
...
         }
         params.put("transformation", transformation.generate());
-        params.put("tag", tag);
+        params.put("tag", options.get("tag"));
+        if (options.containsKey("urls")) {
+            params.put("urls", Arrays.asList((String[]) options.get("urls")));
+        }
         params.put("notification_url", (String) options.get("notification_url"));
         params.put("async", ObjectUtils.asBoolean(options.get("async"), false).toString());
+        params.put("mode", options.get("mode"));
```
Comment (asisayag2) [inline cloudinary-core/src/main/java/com/cloudinary/Uploader.java:301]: @nitzanj , where this `mode` comes from?

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java [hunk 2]
```diff
@@ -282,13 +292,28 @@ public Map generateSprite(String tag, Map options) throws IOException {
...
+
         return callApi("sprite", params, options, null);
     }
 
+    public Map multi(String[] urls, Map options) throws IOException {
+        options.put("urls", urls);
```
Comment (asisayag2) [inline cloudinary-core/src/main/java/com/cloudinary/Uploader.java:307]: Don't you check for array size >=1

File: cloudinary-test-common/src/main/java/com/cloudinary/test/AbstractUploaderTest.java [hunk 3]
```diff
@@ -268,13 +268,19 @@ public void testImageUploadTag() {
...
-        cloudinary.uploader().upload(SRC_TEST_IMAGE, asMap("tags", new String[]{sprite_test_tag, SDK_TEST_TAG, UPLOADER_TAG}, "public_id", "sprite_test_tag_1" + SUFFIX));
-        cloudinary.uploader().upload(SRC_TEST_IMAGE, asMap("tags", new String[]{sprite_test_tag, SDK_TEST_TAG, UPLOADER_TAG}, "public_id", "sprite_test_tag_2" + SUFFIX));
-        Map result = cloudinary.uploader().generateSprite(sprite_test_tag, asMap("tags", Arrays.asList(SDK_TEST_TAG, UPLOADER_TAG)));
+        Map uploadResult1 = cloudinary.uploader().upload(SRC_TEST_IMAGE, asMap("tags", new String[]{sprite_test_tag, SDK_TEST_TAG, UPLOADER_TAG}, "public_id", "sprite_test_tag_1" + SUFFIX));
+        Map uploadResult2 = cloudinary.uploader().upload(SRC_TEST_IMAGE, asMap("tags", new String[]{sprite_test_tag, SDK_TEST_TAG, UPLOADER_TAG}, "public_id", "sprite_test_tag_2" + SUFFIX));
+
+        String[] urls = new String[]{uploadResult1.get("url").toString(), uploadResult2.get("url").toString()};
+
+        Map result = cloudinary.uploader().generateSprite(urls, asMap("tags", Arrays.asList(SDK_TEST_TAG, UPLOADER_TAG)));
```
Comment (asisayag2) [inline cloudinary-test-common/src/main/java/com/cloudinary/test/AbstractUploaderTest.java:279]: Why do we need the `tags` option?

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java [hunk 4]
```diff
@@ -265,9 +265,19 @@ public Map generate_sprite(String tag, Map options) throws IOException {
     }
 
     public Map generateSprite(String tag, Map options) throws IOException {
+        options.put("tag", tag);
+        return generateSprite(options);
+    }
+
+    public Map generateSprite(String[] urls, Map options) throws IOException {
```
Comment (asisayag2) [inline cloudinary-core/src/main/java/com/cloudinary/Uploader.java:272]: What happens if I send both `urls` and `tag`?

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java (hunk 1 above)
Comment (nitzanj) [reply cloudinary-core/src/main/java/com/cloudinary/Uploader.java:301]: You can pass mode=download to both these methods. It doesn't really work that's why it's not documented and not tested. 
The next commit will make more sense of it (As I said this PR is here but not to be merged until discussing the next commit about the download).

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java (hunk 2 above)
Comment (nitzanj) [reply cloudinary-core/src/main/java/com/cloudinary/Uploader.java:307]: We don't run validations at all, we can start if you want but that's not even the tip of the iceberg...

File: cloudinary-test-common/src/main/java/com/cloudinary/test/AbstractUploaderTest.java (hunk 3 above)
Comment (nitzanj) [reply cloudinary-test-common/src/main/java/com/cloudinary/test/AbstractUploaderTest.java:279]: `tags` here just add these tag to the generated sprite (it's not the tag to filter by). We tag anything we can in the tests to better track issues.

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java (hunk 4 above)
Comment (nitzanj) [reply cloudinary-core/src/main/java/com/cloudinary/Uploader.java:272]: See my comment below about validations. Regardless, when calling one of the two public method you pass a parameter directly (not through the map). In order to send both you'd have to manually add the other one to the map - going down this path it's pretty easy to break all the requests.

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java (hunk 1 above)
Comment (asisayag2) [reply cloudinary-core/src/main/java/com/cloudinary/Uploader.java:301]: Let's remove from this PR and add to the relevant one.

File: cloudinary-core/src/main/java/com/cloudinary/Uploader.java (hunk 2 above)
Comment (asisayag2) [reply cloudinary-core/src/main/java/com/cloudinary/Uploader.java:307]: I know we usually don't, and I wouldn't want to retroactively cover all, but in new code, if a validation is calling out + it's easy to validate + it's something that has very low chances of being changed in the BE (we can assume it won't expect an empty array), then why not?

**Pull request merged**